File: android/src/test/kotlin/com/dkmads/dmp/DemographicsTest.kt
package com.dkmads.dmp

import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Test
import java.time.LocalDate

class DemographicsTest {
    private val asOf = LocalDate.of(2026, 6, 18)

    @Test
    fun ageRangeFromDateOfBirth_mapsToCanonicalBucket() {
        assertEquals("25-34", Demographics.ageRangeFromDateOfBirth(LocalDate.of(1992, 3, 15), asOf))
    }

    @Test
    fun ageRangeFromAge_under18_returnsUnknown() {
        assertEquals("unknown", Demographics.ageRangeFromAge(17))
    }

    @Test
    fun ageRangeFromDateOfBirth_string_invalidReturnsNull() {
        assertNull(Demographics.ageRangeFromDateOfBirth("not-a-date", asOf))
    }

    @Test
    fun normalizeAgeRange_rejectsNonCanonical() {
        assertNull(Demographics.normalizeAgeRange("19-26"))
        assertEquals("25-34", Demographics.normalizeAgeRange("25-34"))
    }

    @Test
    fun normalizeGender_lowercasesAndValidates() {
        assertEquals("male", Demographics.normalizeGender("Male"))
        assertNull(Demographics.normalizeGender("nonbinary"))
    }

    @Test
    fun demographicKeys_matchSharedContract() {
        assertEquals("demographic.age_range", Demographics.DEMOGRAPHIC_AGE_RANGE_KEY)
        assertEquals("demographic.gender", Demographics.DEMOGRAPHIC_GENDER_KEY)
    }
}

File: flutter/test/demographics_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:dkmads_dmp/demographics.dart';

void main() {
  final asOf = DateTime(2026, 6, 18);

  test('ageRangeFromDateOfBirth maps to canonical bucket', () {
    expect(
      ageRangeFromDateOfBirth(DateTime(1992, 3, 15), asOf),
      '25-34',
    );
  });

  test('ageRangeFromAge under 18 returns unknown', () {
    expect(ageRangeFromAge(17), 'unknown');
  });

  test('normalizeAgeRange rejects non-canonical values', () {
    expect(normalizeAgeRange('19-26'), isNull);
    expect(normalizeAgeRange('25-34'), '25-34');
  });

  test('normalizeGender lowercases and validates', () {
    expect(normalizeGender('Male'), 'male');
    expect(normalizeGender('nonbinary'), isNull);
  });

  test('demographic keys match shared contract', () {
    expect(demographicAgeRangeKey, 'demographic.age_range');
    expect(demographicGenderKey, 'demographic.gender');
  });
}
